R/modspec.R

```diff
@@ -631,9 +631,10 @@ scrape_and_call <- function(x,env,pass,...) {
 }
 
 dump_opts <- function(x,env,block,...) {
-  hasopt <- unique(c(grep(">>", x, fixed=TRUE),grep("@", x, fixed=TRUE))) 
+  hasopt <- unique(c(grep(">>", x, fixed = TRUE), grep("@", x, fixed = TRUE))) 
+  if(length(hasopt)==0) return(x)
+  hasopt <- grep("^\\s*(>>|@)", x[hasopt], perl = TRUE)
   if(length(hasopt)==0) return(x)
-  hasopt <- grep("^\\s*(>>|@)", x[hasopt], perl=TRUE)
   x[-hasopt]
```

Comment:
The PR description mentions:

> dump_opts had to be refactored to allow an @ sign in quotes passed as run

... but, looking at these changes, it's not popping at to me what change accomplished that.

And looking over this function, I'm getting turned around. I *think* it's supposed to be removing "\>\> ..." and "@ ..." values, but I don't see how it does that. It looks wrong to me that it calculates the `hasopt` indices from the *subsetted* `x` but then uses `-hasopt` to index the full `x`.

Here's what I was playing around (moving the unused arguments):

<details>
<summary>snippet</summary>

``` r
dump_opts_old <- function(x) {
  hasopt <- unique(c(grep(">>", x, fixed=TRUE),grep("@", x, fixed=TRUE)))
  if(length(hasopt)==0) return(x)
  hasopt <- grep("^\\s*(>>|@)", x[hasopt], perl=TRUE)
  x[-hasopt]
}

dump_opts <- function(x) {
  hasopt <- unique(c(grep(">>", x, fixed = TRUE), grep("@", x, fixed = TRUE)))
  if(length(hasopt)==0) return(x)
  hasopt <- grep("^\\s*(>>|@)", x[hasopt], perl = TRUE)
  if(length(hasopt)==0) return(x)
  x[-hasopt]
}

dump_opts_old(c("a", "b"))
#> [1] "a" "b"
dump_opts(c("a", "b"))
#> [1] "a" "b"

dump_opts_old(c("a", "@b"))
#> [1] "@b"
dump_opts(c("a", "@b"))
#> [1] "@b"

dump_opts_old(c("a", "@b", ">>c"))
#> [1] ">>c"
dump_opts(c("a", "@b", ">>c"))
#> [1] ">>c"

dump_opts_old(c("a", "@b", "c"))
#> [1] "@b" "c"
dump_opts(c("a", "@b", "c"))
#> [1] "@b" "c"
```
</details>

Reply:
Here's reprex

``` r
code <- c('root = "@cppstem"', 'run = 1002')

dump_opts_old <- function(x) {
  hasopt <- unique(c(grep(">>", x, fixed=TRUE),grep("@", x, fixed=TRUE)))
  if(length(hasopt)==0) return(x)
  hasopt <- grep("^\\s*(>>|@)", x[hasopt], perl=TRUE)
  x[-hasopt]
}

dump_opts <- function(x) {
  hasopt <- unique(c(grep(">>", x, fixed = TRUE), grep("@", x, fixed = TRUE)))
  if(length(hasopt)==0) return(x)
  hasopt <- grep("^\\s*(>>|@)", x[hasopt], perl = TRUE)
  if(length(hasopt)==0) return(x)
  x[-hasopt]
}

dump_opts(code)
#> [1] "root = \"@cppstem\"" "run = 1002"

dump_opts_old(code)
#> character(0)
```

<sup>Created on 2023-01-11 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Yes; I see it now. Will fix ... need to get some project work out right now. Thanks for picking it up.

Reply:
Removed the over-complexity in `dump_opts()` and added a test. Thanks.